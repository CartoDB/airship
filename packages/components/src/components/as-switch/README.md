# as-switch

It is a two state button that can be on or off. This component issues a "changed" event each time the user presses it, letting it know if it is on or off.

## Examples

### Simple 

```html
<as-switch id="switch-0" name="switch" title="switch"></as-switch>
```

###Â Checked

```html
<as-switch checked id="switch-1" name="switch-1" title="switch-1"></as-switch>
```


### Disabled
```html
<as-switch disabled id="switch-2" name="switch-2" title="switch-2"></as-switch>
```


## Props

### **checked**: boolean
The presence of this boolean attribute controls whether the switch is checked by default or not.

```code
lang: html
---
<as-switch checked ></as-switch>
```
```code
lang: javascript
---
switch.checked = true;
```

### **disabled**: boolean
The presence of this boolean attribute controls when the switch is enabled.

```code
lang: html
---
<as-switch disabled ></as-switch>
```
```code
lang: javascript
---
switch.disabled = true;
```

### **name**: boolean
The name attribute specifies a name for the element. This name attribute can be used to reference the element in a JavaScript.

```code
lang: html
---
<as-switch name="toggle-layer" ></as-switch>
```
```code
lang: javascript
---
switch.name = 'toggle-layer';
```


## Events

### **change** `event.detail = boolean`

Triggered by a enabled switch when the user clicks on it.