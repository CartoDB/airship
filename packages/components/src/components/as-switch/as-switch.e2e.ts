import { EventSpy } from '@stencil/core/dist/declarations';
import { E2EPage, newE2EPage } from '@stencil/core/testing';

describe('as-switch', async () => {
  describe('Rendering', async () => {
    let page: E2EPage;

    beforeEach(async () => {
      page = await newE2EPage({ html: `<as-switch id="switch" label="Switch Label"></as-switch>` });
    });

    it('should render label if prop is present', async () => {
      // TODO: Test
      const element = await page.find('as-switch');
      expect(element.outerHTML).toMatchSnapshot();
    });
  });

  describe('when the component is enabled', () => {
    let page: E2EPage;

    beforeEach(async () => {
      page = await newE2EPage({ html: `<as-switch id="switch"></as-switch>` });
    });

    it('should emit an event with current status when component toggles false->true', async () => {
      const changeSpy = await page.spyOnEvent('change');

      await page.click('as-switch');
      await page.waitForChanges();

      expect(changeSpy).toHaveReceivedEventDetail(true);
    });

    it('should emit an event with current status when component toggles false->true->false->true', async () => {
      const changeSpy: EventSpy = await page.spyOnEvent('change');

      const element = await page.find('as-switch');
      element.click();
      element.click();
      element.click();
      await page.waitForChanges();

      expect(changeSpy).toHaveReceivedEventTimes(3);
      expect(changeSpy.events[0].detail).toBe(true);
      expect(changeSpy.events[1].detail).toBe(false);
      expect(changeSpy.events[2].detail).toBe(true);
    });
  });

  describe('when the component is disabled', () => {
    let page: E2EPage;

    beforeEach(async () => {
      page = await newE2EPage({ html: `<as-switch disabled id="switch"></as-switch>` });
    });

    it('should NOT emit an event with current status when component toggles', async () => {
      const changeSpy: EventSpy = await page.spyOnEvent('change');

      await page.click('as-switch');
      await page.waitForChanges();

      expect(changeSpy).not.toHaveReceivedEvent();
    });
  });

  describe('when the component is checked', () => {
    let page: E2EPage;

    beforeEach(async () => {
      page = await newE2EPage({ html: `<as-switch checked id="switch"></as-switch>` });
    });

    it('should return event (FALSE) when the element is clicked for the first time', async () => {
      const changeSpy: EventSpy = await page.spyOnEvent('change');

      await page.click('as-switch');
      await page.waitForChanges();

      expect(changeSpy).toHaveReceivedEventDetail(false);
    });
  });
});
