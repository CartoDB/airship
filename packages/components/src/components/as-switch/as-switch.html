<!DOCTYPE html>
<html dir="ltr" lang="en">

<head>
  <base href="/">
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=5.0">
  <title>Airship</title>
  <script src="/build/airship.js"></script>
  <link rel="stylesheet" type="text/css" href="/dist/airship.css">

  <style>
    body {
      margin: 10px;
    }

    ul {
      margin: 0;
      padding: 0;
      list-style-type: none;
    }

    .with-label {
      margin-top: 10px;
    }
  </style>
</head>

<body>
  <as-switch id="switch-0" name="switch" title="switch"></as-switch>
  <as-switch checked id="switch-1" name="switch-1" title="switch-1"></as-switch>
  <as-switch disabled id="switch-2" name="switch-2" title="switch-2"></as-switch>

  <div class="with-label">
    <as-switch checked id="switch-3" name="switch-3" title="switch-3" label="Private Room"></as-switch>
  </div>

  <script>
    document.querySelector('#switch-0').addEventListener('change', console.warn);

    document.querySelector('#switch-1').addEventListener('change', console.warn);

  </script>
</body>

</html>
