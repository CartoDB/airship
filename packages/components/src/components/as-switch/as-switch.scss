@import '../../../../styles/src/core/_dev';

as-switch {
  --as-switch--color--background-default: var(--as--color--type-03);
  --as-switch--color--background-disabled: var(--as--color--ui-02);
  --as-switch--color--background-checked: var(--as--color--primary);
  --as-switch--color--background-active: var(--as--color--complementary);
  --as-switch--color--thumb: var(--as--color--ui-01);

  display: inline-block;
  position: relative;

  .as-switch {
    display: flex;
    align-items: center;
  }

  .as-switch__label {
    margin: 0 0 0 get-spacing(1);
  }

  .as-switch__thumb {
    position: absolute;
    top: 50%;
    left: get-spacing(1);
    transform: translate3d(0, -50%, 0);
    transition: opacity 0.2s ease;
    opacity: 0;
    color: var(--as-switch--color--thumb);
    pointer-events: none;
  }

  .as-switch__input {
    position: absolute;
    width: 1px;
    height: 1px;
    margin: -1px;
    padding: 0;
    overflow: hidden;
    clip: rect(0 0 0 0);
    border: 0;
  }

  .as-switch__input + .as-switch__toggle {
    display: block;
    position: relative;
    width: get-spacing(8);
    height: get-spacing(4);
    transition: all 0.4s ease;
    border-radius: get-spacing(10);
    outline: 0;
    background: var(--as-switch--color--background-default);
    cursor: pointer;
    user-select: none;
  }

  .as-switch__toggle:hover {
    background: var(--as-switch--color--background-active);
  }

  .as-switch__input + .as-switch__toggle::after {
    content: '';
    display: block;
    position: relative;
    top: 2px;
    width: get-spacing(3);
    height: get-spacing(3);
    transform: translateX(2px);
    transition: all 0.2s ease;
    border-radius: 50%;
    background: var(--as-switch--color--thumb);
  }

  .as-switch__input[disabled] + .as-switch__toggle {
    background: var(--as-switch--color--background-disabled);
    box-shadow: inset 0 0 0 1px var(--as--color--ui-03, #{$color-ui-03});
    cursor: not-allowed;
  }

  .as-switch__input[disabled] + .as-switch__toggle::after {
    background: var(--as--color--ui-04, #{$color-ui-04});
  }

  .as-switch__input:checked + .as-switch__toggle,
  .as-switch__input:checked + .as-switch__toggle:hover {
    background: var(--as--color--primary, #{$color-primary});
  }

  .as-switch__input:checked + .as-switch__toggle::after {
    transform: translateX(18px);
  }

  .as-switch__input:checked ~ svg {
    opacity: 1;
  }
}
