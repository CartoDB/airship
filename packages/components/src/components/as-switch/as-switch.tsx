import { Component, Element, Event, EventEmitter, Prop } from '@stencil/core';

@Component({
  shadow: false,
  styleUrl: './as-switch.scss',
  tag: 'as-switch',
})

export class Switch {
  /**
   * HTMLElement wraper for the switch component
   *
   * @type {HTMLElement}
   * @memberof Switch
   */
  @Element() public el: HTMLElement;

  /**
   * Boolean flag to control if the input is checked or not
   *
   * @type{boolean}
   * @memberof Switch
   */
  @Prop({ mutable: true, reflectToAttr: true }) public checked: boolean;

  /**
   * Boolean flag to control when the switch is disabled or not
   *
   * @type{boolean}
   * @memberof Switch
   */
  @Prop({ reflectToAttr: true }) public disabled: boolean;

  /**
   * The input name
   *
   * @type{string}
   * @memberof Switch
   */
  @Prop({ reflectToAttr: true }) public name: string;

  /**
   * Event triggered by a enabled Switch component when the user clicks on it.
   *
   * @type{boolean}
   * @memberof Switch
   */
  @Event() public change: EventEmitter;

  /**
   * When the component is attached to the DOM bind the onClick function
   */
  public componentDidLoad() {
    this.el.addEventListener('click', this._onClick.bind(this));
  }

  /**
   * Fire a 'change' event with a boolean parameter if the user clicks on an enabled component
   */
  public _onClick() {
    const input = this.el.querySelector('input');
    if (!this.disabled) {
      input.checked = !input.checked;
      this.checked = input.checked;
      this.change.emit(input.checked);
    }
  }

  public render() {
    return [
      <input
        class='as-switch__input'
        checked={this.checked}
        disabled={this.disabled}
        role='switch'
        id={this.el.id}
        name={this.name}
        type='checkbox'
        title={this.el.title}
      />,
      <label class='as-switch__toggle' htmlFor='switch_0'></label>,
      <svg class='as-switch__thumb' width='10' height='8' xmlns='http://www.w3.org/2000/svg'>
        <path
          fill='currentColor'
          // tslint:disable-next-line
          d='M3.315 7.858L.133 4.441a.506.506 0 0 1 0-.683l.643-.684a.437.437 0 0 1 .642 0l2.219 2.393L8.58.141a.437.437 0 0 1 .643 0l.643.683a.504.504 0 0 1 0 .683l-5.91 6.35a.437.437 0 0 1-.642 0' />
      </svg>
    ];
  }
}
