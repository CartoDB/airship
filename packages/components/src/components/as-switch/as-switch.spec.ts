import { TestWindow } from '@stencil/core/dist/testing';
import { Switch } from './as-switch';

describe.only('as-switch', () => {
  it('should build', () => {
    expect(new Switch()).toBeTruthy();
  });

  describe('Rendering', () => {
    it('should render properly', async () => {
      const switchElement = await _prepareTest('<as-switch id="switch"></as-switch>');

      expect(switchElement).toMatchSnapshot();
    });

    it('should render label if prop is present', async () => {
      const switchElement = await _prepareTest('<as-switch id="switch" label="Switch Label"></as-switch>');

      expect(switchElement).toMatchSnapshot();
    });
  });

  describe('when the component is enabled', () => {
    let element: HTMLAsSwitchElement;

    beforeEach(async () => {
      element = await _prepareTest('<as-switch id="switch"></as-switch>');
    });

    it('should emit an event with current status when component toggles false->true', async () => {
      const spy = _spyOnElement(element);

      element.click();

      expect(spy).toHaveBeenCalled();
      expect(spy.mock.calls[0][0].detail).toBe(true);
    });

    it('should emit an event with current status when component toggles false->true->false->true', async () => {
      const spy = _spyOnElement(element);


      element.click();
      element.click();
      element.click();

      expect(spy).toHaveBeenCalled();
      expect(spy.mock.calls[0][0].detail).toBe(true);
      expect(spy.mock.calls[1][0].detail).toBe(false);
      expect(spy.mock.calls[2][0].detail).toBe(true);
    });
  });

  describe('when the component is disabled', () => {
    let element: HTMLAsSwitchElement;

    beforeEach(async () => {
      element = await _prepareTest('<as-switch disabled id="switch"></as-switch>');
    });

    it('should NOT emit an event with current status when component toggles', async () => {
      const spy = _spyOnElement(element);

      element.click();

      expect(spy).not.toHaveBeenCalled();
    });
  });

  describe('when the component is checked', () => {
    let element: HTMLAsSwitchElement;

    beforeEach(async () => {
      element = await _prepareTest('<as-switch checked id="switch"></as-switch>');
    });

    it('should return event (FALSE) when the element is clicked for the first time', async () => {
      const spy = _spyOnElement(element);
      element.click();

      expect(spy).toHaveBeenCalled();
      expect(spy.mock.calls[0][0].detail).toBe(false);
    });
  });
});

async function _prepareTest(html) {
  const tw = new TestWindow();
  const e = await tw.load({
    components: [Switch],
    html,
  });
  await tw.flush();
  return e;
}

function _spyOnElement(element) {
  const spy = jest.fn();
  element.addEventListener('change', spy);
  return spy;
}
