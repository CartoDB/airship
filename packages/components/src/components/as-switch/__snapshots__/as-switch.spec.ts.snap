// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`as-switch Rendering should render label if prop is present 1`] = `
<as-switch
  class="hydrated"
  id="switch"
  label="Switch Label"
>
  
  <div
    class="as-switch"
  >
    <div
      class="as-switch__element"
    >
      <input
        class="as-switch__input"
        id="switch"
        role="switch"
        title=""
        type="checkbox"
      />
      <label
        class="as-switch__toggle"
        for="switch"
      />
      <svg
        class="as-switch__thumb"
        height="8"
        width="10"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M3.315 7.858L.133 4.441a.506.506 0 0 1 0-.683l.643-.684a.437.437 0 0 1 .642 0l2.219 2.393L8.58.141a.437.437 0 0 1 .643 0l.643.683a.504.504 0 0 1 0 .683l-5.91 6.35a.437.437 0 0 1-.642 0"
          fill="currentColor"
        />
      </svg>
    </div>
    <label
      class="as-switch__label as-body"
    >
      Switch Label
    </label>
  </div>
</as-switch>
`;

exports[`as-switch Rendering should render properly 1`] = `
<as-switch
  class="hydrated"
  id="switch"
>
  
  <input
    class="as-switch__input"
    id="switch"
    role="switch"
    title=""
    type="checkbox"
  />
  <label
    class="as-switch__toggle"
    for="switch"
  />
  <svg
    class="as-switch__thumb"
    height="8"
    width="10"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M3.315 7.858L.133 4.441a.506.506 0 0 1 0-.683l.643-.684a.437.437 0 0 1 .642 0l2.219 2.393L8.58.141a.437.437 0 0 1 .643 0l.643.683a.504.504 0 0 1 0 .683l-5.91 6.35a.437.437 0 0 1-.642 0"
      fill="currentColor"
    />
  </svg>
</as-switch>
`;
